File: core/src/main/scala/io/github/mathter/zi/dsl/Dsl.scala
package io.github.mathter.zi.dsl

trait Dsl
  extends LiteralDsl
    with ListDsl
    with OriginDsl
    with ResultDsl
    with PathMapDsl
    with ConstDsl {

  def nothing[T]: Source[T]

  def unit[T](source: Source[T]): Source[T] = source

  def `if`[T](condition: Source[Boolean]): If[T]
}

given [T](using dsl: Dsl): Conversion[T, Source[T]] with {
  override def apply(x: T): Source[T] = dsl.literal(x)
}

implicit class ListOps[T](x: Source[T]) {
  def list: Source[List[T]] = x.dsl.list(x)
}

File: jcore/src/main/java/io/github/mathter/jzi/dsl/Dsl.java
package io.github.mathter.jzi.dsl;

import io.github.mathter.zi.dsl.LiteralDsl;
import io.github.mathter.zi.dsl.OriginDsl;
import io.github.mathter.zi.dsl.ResultDsl;
import io.github.mathter.zi.dsl.Source;

import java.math.BigDecimal;
import java.math.BigInteger;
import java.util.function.Supplier;

public interface Dsl extends
        LiteralDsl,
        OriginDsl,
        ResultDsl {
    public StringSource asStringSource(Source<String> source);

    public NumberSource<Byte> asByteSource(Source<Byte> source);

    public NumberSource<Short> asShortSource(Source<Short> source);

    public NumberSource<Integer> asIntSource(Source<Integer> source);

    public NumberSource<Long> asLongSource(Source<Long> source);

    public NumberSource<Float> asFloatSource(Source<Float> source);

    public NumberSource<Double> asDoubleSource(Source<Double> source);

    public NumberSource<BigInteger> asBigIntegerSource(Source<BigInteger> source);

    public NumberSource<BigDecimal> asBigDecimalSource(Source<BigDecimal> source);

    public NumberSource<Byte> literal(Byte value);

    public NumberSource<Short> literal(Short value);

    public NumberSource<Integer> literal(Integer value);

    public NumberSource<Long> literal(Long value);

    public NumberSource<Float> literal(Float value);

    public NumberSource<Double> literal(Double value);

    public NumberSource<BigInteger> literal(BigInteger value);

    public NumberSource<BigDecimal> literal(BigDecimal value);

    public <T extends Number> NumberSource<T> numberLiteral(Supplier<T> supplier);

    public StringSource literal(String value);

    public StringSource stringLiteral(Supplier<String> supplier);

    public BooleanSource literal(Boolean literal);

    public BooleanSource booleanLiteral(Supplier<Boolean> supplier);
}
